main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -275,14 +277,47 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
               ///
               Validation.mapN(annVal) {
                 case as =>
-                  (TypedAst.Def(doc, as, mod, sym, tparams, fparams, exp, sc, inferredScheme, declaredEff, loc), subst)
+                  (TypedAbstractDefinition(as, tparams, fparams, exp, sc, inferredScheme, declaredEff, loc), subst)
               }
 
             case Err(e) => Validation.Failure(LazyList(e))
           }
       }
   }
 
+  /**
+    * Infers the type of the given signature `sig0`.
+    */
+  private def typeCheckSig(sig0: ResolvedAst.Sig, assumedTconstrs: List[Ast.TypeConstraint], root: ResolvedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[(TypedAst.Sig, Substitution), TypeError] = sig0 match {
+    // Case 1: signature without implementation
+    case ResolvedAst.Sig(doc, ann0, mod, sym, tparams0, fparams0, None, sc, eff, loc) =>
+      val tparams = getTypeParams(tparams0)
+      val fparams = getFormalParams(fparams0, Substitution.empty)
+      for {
+        ann <- visitAnnotations(ann0, root)
+      } yield (TypedAst.Sig(doc, ann, mod, sym, tparams, fparams, None, sc, None, eff, loc), Substitution.empty)
+    // Case 2: signature with implementation
+    case ResolvedAst.Sig(doc0, ann0, mod0, sym0, tparams0, fparams0, Some(exp0), sc0, eff0, loc0) =>
+      val abstractDef = ResolvedAbstractDefinition(ann0, tparams0, fparams0, exp0, sc0, eff0, loc0)
+      Validation.mapN(typeCheckAbstractDef(abstractDef, assumedTconstrs, isMain = false, root, classEnv)) {
+        case (TypedAbstractDefinition(ann, tparams, fparams, exp, declaredScheme, inferredScheme, eff, loc), subst) =>
+          (TypedAst.Sig(doc0, ann, mod0, sym0, tparams, fparams, Some(exp), declaredScheme, Some(inferredScheme), eff, loc), subst)
+      }
+  }
+
+  /**
+    * Infers the type of the given definition `def0`.
+    */
+  private def typeCheckDef(def0: ResolvedAst.Def, assumedTconstrs: List[Ast.TypeConstraint], root: ResolvedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[(TypedAst.Def, Substitution), TypeError] = def0 match {
+    case ResolvedAst.Def(doc0, ann0, mod0, sym0, tparams0, fparams0, exp0, sc0, eff0, loc0) =>
+      val abstractDef = ResolvedAbstractDefinition(ann0, tparams0, fparams0, exp0, sc0, eff0, loc0)
+
+      Validation.mapN(typeCheckAbstractDef(abstractDef, assumedTconstrs, sym0.isMain, root, classEnv)) {
+        case (TypedAbstractDefinition(ann, tparams, fparams, exp, declaredScheme, inferredScheme, eff, loc), subst) =>
+          (TypedAst.Def(doc0, ann, mod0, sym0, tparams, fparams, exp, declaredScheme, inferredScheme, eff, loc), subst)
+      }
+  }
+
```

Comment:
I am a little concerned with all these changes to the Typer. I assume there is a reason that we need to distinguish defs and sigs (with impl). I can guess at two important differences: available type class constraints and whether there is a body at all.

I can sort of see the reason for the structure, I am just not to keen on it.